### PR TITLE
Hard computer opponent

### PR DESCRIPTION
**Pull request by samjones1001**

### CODE REVIEW

File: tests/test_board.py [hunk 1]
```diff
@@ -16,27 +16,18 @@ def test_board_has_nine_spaces_by_default(board):
...
 def test_board_is_not_full_when_no_spaces_are_occupied(board):
     assert board.is_full() is False
 
 
-def test_board_is_not_full_when_only_some_spaces_are_occupied(board):
-    num_of_spaces_minus_one = len(board.get_spaces()) - 1
-    for space in range(num_of_spaces_minus_one):
-        board.place_marker(space, 'x')
+def test_board_is_not_full_when_only_some_spaces_are_occupied():
+    board = Board(['x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', '-'])
```
Comment (amb8l) [inline tests/test_board.py:24]: :+1:

File: tests/test_board.py [hunk 2]
```diff
@@ -46,9 +37,19 @@ def test_a_space_is_available_when_it_exists_and_is_unoccupied(board):
 
 
 def test_a_space_is_unavailable_when_it_exists_but_is_occupied(board):
-    board.place_marker(0, 'x')
+    new_board = board.place_marker(0, 'x')
+
+    assert 0 not in new_board.available_spaces()
```
Comment (amb8l) [inline tests/test_board.py:42]: :+1:

File: tests/test_board.py [hunk 3]
```diff
@@ -57,32 +58,25 @@ def test_returns_all_spaces_when_none_have_been_occupied(board):
     assert len(board.available_spaces()) == number_of_spaces
 
 
-def test_only_returns_unoccupied_spaces_when_some_have_been_occupied(board):
-    board.place_marker(0, 'X')
+def test_does_not_return_occupied_spaces(board):
```
Comment (amb8l) [inline tests/test_board.py:61]: Not sure how this test is any different from `test_a_space_is_unavailable_when_it_exists_but_is_occupied`?

File: tests/test_board.py [hunk 4]
```diff
@@ -57,32 +58,25 @@ def test_returns_all_spaces_when_none_have_been_occupied(board):
...
+    assert 0 not in new_board.available_spaces()
 
 
-def test_returns_no_spaces_when_all_have_been_occupied(board):
-    for space in range(0, len(board.get_spaces())):
-        board.place_marker(space, 'X')
+def test_returns_no_spaces_when_all_have_been_occupied():
+    board = Board(['x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x'])
 
     assert len(board.available_spaces()) == 0
```
Comment (amb8l) [inline tests/test_board.py:70]: Good :+1:

How do you feel about `assert board.available_spaces() == []`?

File: tests/mocks.py [hunk 5]
```diff
@@ -56,54 +52,58 @@ def get_name(self):
...
+        self.play_turn_call_count = 0
+        self.show_game_over_screen_call_count = 0
 
     def get_board_state(self):
         return self._board_state
 
+    def available_spaces(self):
+        return self._available_spaces
+
+    def game_over(self, console):
```
Comment (Kotauror) [inline tests/mocks.py:78]: What was your reason for mocking the whole Game? I'm not sure why functions like `available_spaces` need to be mocked. I feel that mocking the whole game is too much 🤔

File: tests/test_menu.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+from ttt.console_ui.menu import Menu
+from ttt.players.human_player import HumanPlayer
+from ttt.players.simple_computer_player import SimpleComputerPlayer
+from ttt.players.smart_computer_player import SmartComputerPlayer
+from tests.mocks import MockConsole, MockGameRunner
+
+
+def test_starting_a_game_requests_a_welcome_message_be_displayed():
+    runner = MockGameRunner('console')
```
Comment (Kotauror) [inline tests/test_menu.py:9]: This could be placed in a before block.

File: tests/test_menu.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    interface = Menu(MockConsole(input_return='1'))
+    interface.start(runner)
+
+    assert runner.run_call_count == 1
+
+
+def test_user_can_select_a_human_opponent():
+    runner = MockGameRunner('console')
+    interface = Menu(MockConsole(input_return='1'))
+    interface.start(runner)
```
Comment (Kotauror) [inline tests/test_menu.py:28]: as above - there is quite some repetition going on here

File: tests/test_smart_computer_player.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
+from ttt.players.smart_computer_player import SmartComputerPlayer
+from ttt.game.game import Game
+from ttt.game.board import Board
+
+
+def test_will_always_pick_the_center_space_if_available():
+    board = Board(['O','-','-','-','-','-','-','-','-'])
+    player_1 = SmartComputerPlayer('Player 1', 'X')
+    player_2 = SmartComputerPlayer('Player 2', 'O')
+    game = Game(player_1, player_2, board)
```
Comment (Kotauror) [inline tests/test_smart_computer_player.py:10]: Hmm, my issue with having a `MockGame` is that sometimes you use `MockGame` and sometimes `Game`, but you pass it board with certain spacer available. I'm not sure what's the key to understanding where one should use `Game`, not `MockGame` and not the other way around.

**Review by Kotauror — COMMENTED** [review]
I've started reviewing from the oldest PR, hopefully my comments won't be outdated 😅 In case they are, please ignore them :D

File: tests/test_smart_computer_player.py (hunk 8 above)
Comment (samjones1001) [reply tests/test_smart_computer_player.py:10]: Agreed - I think my mocks have got to a position of being a bit confused - I'm planning to take a look tomorrow at streamlining them.

**Pull request merged**